### PR TITLE
Create and use test harness for the Windows kernel (freestanding) (#1289)

### PR DESCRIPTION
**Pull request by ben-craig**
WIP: I'm well aware that this isn't acceptable as is.  This is initial work so that I can try to get concrete feedback earlier in the process.

This adds the capability to test the STL in the Windows kernel.  The expectation is that the Windows kernel will support most of the freestanding facilities, with the likely exception of those involving RTTI and exceptions.

On the cmake side of things, I add a new executable that will do the necessary Windows things to load a kernel driver and run the unit test.  I also create a static lib with shared per-test code.  This includes a custom assert implementation that won't crash the system, but will instead continue executing and report the results up though the executable.  This is basically downgrading the "assert" to an "expect".  In practice, this hasn't caused me any problems so far.

On the lit side of things, I have code scattered all over the place (this is one of the things that I'm hoping to improve and get guidance on).  I have a fair number of linker and compiler flags stuck in the lit code (as opposed to the .lst files).  I currently have driver signing shoe-horned into the lit format class, so that I only sign tests that actually need to run.  I also abuse a force include to inject the "name" of the driver into each .cpp file (this doesn't work for multi TU tests yet).  This force include injection makes it so that I don't need to invoke the compiler an extra time per test.

The certificate used for driver signing is generated at test time, and is self signed.  The certificate has a password that is generated during test time, and then forgotten when the test is done running.  The certificate also has a very quick expiration date (2 days).  All of these together should make the certificate both difficult to attack, and not worth attacking, as it will (at most) give you access to one CI VM for 2 days.

### CODE REVIEW

**Comment by StephanTLavavej** [comment]
@cbezault is planning to clean up the underlying test machinery, which should make your life easier (but will require changes to this PR).

File: tests/libcxx/usual_matrix.lst [hunk 1]
```diff
@@ -3,7 +3,8 @@
 
 RUNALL_INCLUDE ..\universal_prefix.lst
 RUNALL_CROSSLIST
-PM_CL="/EHsc /MTd /std:c++latest /permissive- /FImsvc_stdlib_force_include.h /wd4643"
+# PM_CL="/EHsc /MTd /std:c++latest /permissive- /FImsvc_stdlib_force_include.h /wd4643"
+PM_CL="/kernel /std:c++latest /permissive- /FImsvc_stdlib_force_include.h /wd4643"
 RUNALL_CROSSLIST
 PM_CL="/analyze /Zc:preprocessor"
-PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing"
+# PM_COMPILER="clang-cl" PM_CL="-fno-ms-compatibility -fno-delayed-template-parsing"
```
Comment (ben-craig) [inline tests/libcxx/usual_matrix.lst:10]: This file shouldn't be merged, as it is just a prototype hack.  We will need to figure out what to do with the /EHsc, /MTd, and /kernel flags at some point.

File: tests/utils/kernel/inc/stl_kernel/kernel_test_constants.h [hunk 2]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// This header is intended to be force included into the .cpp under test.
+// KERNEL_TEST_NAME will then be provided as a /D switch and preprocessor
+// concatenated into one big string literal.
```
Comment (ben-craig) [inline tests/utils/kernel/inc/stl_kernel/kernel_test_constants.h:6]: This approach may need to change, depending on whether we can figure out another way to do multi-TU tests or not.

File: tests/utils/kernel/src/stl_kernel/test_decls.h [hunk 3]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#include <ntddk.h>
+
+extern const wchar_t *STL_KERNEL_NT_DEVICE_NAME;
+extern const wchar_t *STL_KERNEL_DOS_DEVICE_NAME;
```
Comment (ben-craig) [inline tests/utils/kernel/src/stl_kernel/test_decls.h:8]: I missed a const on the * here (and in the definition)

File: tests/utils/kernel/src/stl_kernel_loader/install.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+        error_buffer,
+        static_cast<DWORD>(std::size(error_buffer)),
+        nullptr /* va_list arguments */);
+
+    std::string err_text = "Error: ";
+    err_text += routine;
+    err_text += " failed.  GetLastError = " + std::to_string(GetLastError());
+    err_text += "\n";
+    err_text += error_buffer;
```
Comment (ben-craig) [inline tests/utils/kernel/src/stl_kernel_loader/install.cpp:26]: Looking forward to std::format!

File: tests/utils/stl/compiler.py [hunk 5]
```diff
@@ -87,6 +88,11 @@ def _basicCmd(self, source_files: List[Path], out: Path,
             cmd.extend(self.compile_flags)
             cmd.extend(compile_flags)
 
+        if self.is_kernel and mode in (self.CM_PreProcess, self.CM_Compile, self.CM_Default):
+            name = str(out).replace('\\','.').replace(':','.')
+            cmd.extend(['/DKERNEL_TEST_NAME=L"' + name + '"'])
+            # TODO need to sign the linked binary
```
Comment (ben-craig) [inline tests/utils/stl/compiler.py:94]: TODO is TODONE, just in a different file

File: tests/utils/stl/test/config.py [hunk 6]
```diff
@@ -358,6 +405,30 @@ def configure_link_flags(self):
...
+            arch_name = None
+            if self.target_arch.casefold() == 'AMD64'.casefold():
+                arch_name = 'x64'
+            else:
+                arch_name = self.target_arch.lower()
+
+            self.default_compiler.link_flags.extend([
+                '/LIBPATH:' + self.wdk_lib + '/km/' + arch_name,
+                '/IGNORE:4210',
+                '/machine:'+arch_name,
```
Comment (ben-craig) [inline tests/utils/stl/test/config.py:418]: Note: I haven't tested 32-bit kernels at all.

File: tests/utils/stl/test/executor.py [hunk 7]
```diff
@@ -76,12 +76,10 @@ def __init__(self, commandPrefix, chain):
         self.commandPrefix = commandPrefix
         self.chain = chain
 
-    def run(self, exe_path, cmd=None, work_dir='.', file_deps=None, env=None):
-        cmd = cmd or [exe_path]
-        return self.chain.run(exe_path, self.commandPrefix + cmd, work_dir,
+    def run(self, cmd, work_dir='.', file_deps=None, env=None):
+        return self.chain.run([self.commandPrefix] + cmd, work_dir,
                               file_deps, env=env)
```
Comment (ben-craig) [inline tests/utils/stl/test/executor.py:81]: This was just broken.  PostFix looks broken too, but I haven't had a need to use it.

File: tests/utils/stl/test/format.py [hunk 8]
```diff
@@ -31,7 +32,31 @@ class TestStep:
...
+            # sign the binary before running it.
+            cmd = [self.wdk_bin + '/x64/signtool.exe', 'sign',
+                   '/f', self.cert_path,
+                   '/p', self.cert_pass,
+                   self.cmd[0]]
+            out, err, rc = stl.util.executeCommand(cmd, self.work_dir)
+            if rc != 0:
+                self.should_fail = False
+                return (cmd, out, err, rc)
+        return super().run()
```
Comment (ben-craig) [inline tests/utils/stl/test/format.py:59]: As the comment suggests, I don't think too highly of this approach.  A more correct approach may be to add a build step enum that covers signing, or to make the link step also do signing.

**Comment by StephanTLavavej** [comment]
@cbezault says that #1421 is expected to supersede this PR; he's going to give you push access to his fork.

**Comment by cbezault** [comment]
Let's move everything to #1421. (Unless you object)

**Pull request closed without merging**